key.core/src/main/java/de/uka/ilkd/key/speclang/jml/translation/JMLSpecFactory.java

```diff
@@ -1347,19 +1344,76 @@ public void translateJmlAssertCondition(final JmlAssert jmlAssert, final IProgra
-                                                                                         // enclosing
-                                                                                         // contract
+            append(collectLocalVariablesVisibleTo(statement, pm), parameters);
+        return new ProgramVariableCollection(tb.selfVar(pm, pm.getContainerType(), false),
+            paramVars, tb.resultVar(pm, false), tb.excVar(pm, false), atPreVars, termify(atPreVars),
+            Collections.emptyMap(), // should be the pre-state of the enclosing contract
             Collections.emptyMap() // ignore for now
         );
-        var io = new JmlIO(services).context(Context.inMethodWithSelfVar(pm, pv.selfVar));
-        jmlAssert.translateCondition(io, pv);
+    }
+
+    /**
+     * Translates the condition Term of a JmlAssert statement.
+     *
+     * @param jmlAssert the statement to create the condition for
+     * @param pm the enclosing method
+     */
+    public void translateJmlAssertCondition(final JmlAssert jmlAssert, final IProgramMethod pm) {
+        final var pv = createProgramVariablesForStatement(jmlAssert, pm);
+        jmlAssert.translateCondition(new JmlIO(services).context(Context.inMethod(pm, tb)), pv);
+    }
+
+    public @Nullable String checkSetStatementAssignee(Term assignee) {
+        if (assignee.op() instanceof LocationVariable) {
+            var variable = (LocationVariable) assignee.op();
+            if (variable.isGhost()) {
+                return null;
+            } else {
+                return variable + " is not a ghost variable";
+            }
+        } else if (services.getTypeConverter().getHeapLDT().isSelectOp(assignee.op())) {
+            var field = assignee.subs().last();
+            var op = field.op();
+            var split = HeapLDT.trySplitFieldName(op);
+            if (split != null) {
+                var attribute =
+                    services.getJavaInfo().getAttribute(split.attributeName, split.className);
+                if (attribute.isGhost()) {
+                    return null;
+                } else {
+                    return op + " is not a ghost field";
+                }
+            } else if (op.equals(services.getTypeConverter().getHeapLDT().getArr())) {
+                return "Arrays are not writeable using set statements";
+            } else {
+                return op + " is not a class field";
+            }
+        } else {
+            return "Neither a field access nor a local variable access";
+        }
+    }
+
+    /**
+     * Translates a set statement.
+     *
+     * @param statement the set statement
+     * @param pm the enclosing method
+     */
+    public void translateSetStatement(final SetStatement statement, final IProgramMethod pm)
+            throws SLTranslationException {
+        final var pv = createProgramVariablesForStatement(statement, pm);
+        var setStatementContext = statement.takeParserContext();
```

Comment:
type? pls make explicit

Reply:
I'm reluctant in adding these types since they are the generated types coming from the JmlParser and their names convey almost no meaning.